File: archivelib-sys2/c-lib/src/r_expand/get_next_item.cpp

#include <cassert>

#include "support/expand.hpp"

#include "r_expand.hpp"

uint16_t RExpand::get_next_item() {
  uint16_t run_length276, _283;
  if (data->items_until_next_header == 0) {
    // This is the first 2 bytes in the file, and it represents the number of
    // calls that this header can handle. It's not exactly the number of bytes
    // because we read a variable number of bits per call.
    data->items_until_next_header = get_bits(16);
    DE;
    fn253(CONST_N145_IS_19, CONST_N147_IS_5, 3);
    DE;
    fn255();
    DE;
    fn253(CONST_N142_IS_15, CONST_N540_IS_5, -1);
    DE;
    if (mStatus < 0)
      return 0;
  }
  data->items_until_next_header--;
  run_length276 = data->dat_arr240[data->bits182 >> 4];
  // run_length276 <= 0xFF are the uncompressed bits.
  // 0x100 <= run_length276 <= 0x1FE are runs (run_length276 - 0x100 + 3) bits
  // long
  if (run_length276 >= CONST_N141_IS_511) {
    // No test cases exercise this condition.
    printf("\n");
    WRITE_BITS(std::cout, "", data->bits182);
    printf("\n");
    _283 = 1U << 3;
    do {
      printf("                _283: %#4x\n", _283);
      printf("data->bits182 & _283: %#4x\n", data->bits182 & _283);
      printf("       run_length276: %#x\n", run_length276);
      if (data->bits182 & _283)
        run_length276 = data->dat_arr190[run_length276];
      else
        run_length276 = data->dat_arr189[run_length276];
      _283 >>= 1;
    } while (run_length276 >= CONST_N141_IS_511);
    printf("\n");
    abort();
  }
  read_bits(data->dat_arr180[run_length276]);
  return run_length276;
}

File: src/expand/get_next_item.rs
use crate::consts::{
  CONST_N141_IS_511, CONST_N142_IS_15, CONST_N145_IS_19, CONST_N147_IS_5, CONST_N540_IS_5,
};
use crate::expand::{RExpandData, Result};
use crate::support::BitRead;
use std::io::Write;

impl<R: BitRead, W: Write> RExpandData<R, W> {
  pub fn get_next_item(&mut self) -> Result<u16> {
    if self.items_until_next_header == 0 {
      // This is the first 2 bytes in the file, and it represents the number of
      // calls that this header can handle. It's not exactly the number of bytes
      // because we read a variable number of bits per call.
      self.items_until_next_header = self.get_bits(16)? as usize;
      self.fn253(i16::try_from(CONST_N145_IS_19).unwrap(), i16::try_from(CONST_N147_IS_5).unwrap(), 3)?;
      self.fn255()?;
      self.fn253(i16::try_from(CONST_N142_IS_15).unwrap(), i16::try_from(CONST_N540_IS_5).unwrap(), -1)?;
      // println!("dat_arr180: {:X?}", self.dat_arr180);
      // println!("dat_arr181: {:X?}", self.dat_arr181);
      // println!("dat_arr189: {:X?}", self.dat_arr189);
      // println!("dat_arr190: {:X?}", self.dat_arr190);
      // println!("dat_arr240: {:X?}", self.dat_arr240);
      // println!("dat_arr241: {:X?}", self.dat_arr241);
    }
    self.items_until_next_header -= 1;
    let mut run_length276 = self.dat_arr240[(self.bits182 >> 4) as usize];
    // run_length276 <= 0xFF are the uncompressed bits.
    // 0x100 <= run_length276 <= 0x1FE are runs (run_length276 - 0x100 + 3) bits
    // long
    if usize::try_from(run_length276).unwrap() >= CONST_N141_IS_511 {
      // No test cases exercise this condition.
      let mut var283 = 1_u16 << 3;
      loop {
        if 0 != self.bits182 & var283 {
          run_length276 = self.dat_arr190[usize::try_from(run_length276).unwrap()]
        } else {
          run_length276 = self.dat_arr189[usize::try_from(run_length276).unwrap()]
        }
        var283 >>= 1;
        if (usize::try_from(run_length276).unwrap()) < CONST_N141_IS_511 {
          break;
        }
      }
    }
    let bits = i16::from(self.dat_arr180[usize::try_from(run_length276).unwrap()]);
    self.read_bits(bits)?;
    Ok(run_length276)
  }
}
